### PR TITLE
OmniSci Join dwarf

### PR DESCRIPTION
**Pull request by bagrorg**

### CODE REVIEW

File: common/dpcpp/omnisci_hashtable.hpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  }
+
+  void build_table() {
+    q.submit([&](sycl::handler &cgh) {
+       ht_fields loc_f = *f;
+       auto ht = hash_table.get_access(cgh);
+       auto ks = keys.get_access(cgh);
+
+       cgh.parallel_for(loc_f.keys_size, [=](auto i) {
+         auto h = loc_f.hasher(ks[i]);
```
[the remark this answers is not in the document]
Comment (bagrorg) [reply common/dpcpp/omnisci_hashtable.hpp:62]: Yes
Previousle, we made sure that SYCL distributes the work better than if we used `for` inside `parallel_for`

File: common/dpcpp/omnisci_hashtable.hpp [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+      : q(q), keys(keys_vec), hash_table(ht_size), count_buffer(ht_size),
+        position_buffer(ht_size), id_buffer(keys_vec.size()) {
+
+    ht_fields tmp_f = {empty_key, ht_size,     distinct_size, keys_vec.size(),
+
+                       ht_size,   ht_size * 2, ht_size * 3,
+
+                       hasher};
+
+    f = std::make_unique<ht_fields>(tmp_f);
```
[the remark this answers is not in the document]
Comment (bagrorg) [reply common/dpcpp/omnisci_hashtable.hpp]: I used pimp to reduce accessing code in submits, e.g. [here](https://github.com/bagrorg/dwarf_bench/blob/snd_join_omnisci/common/dpcpp/omnisci_hashtable.hpp#L35)

File: common/dpcpp/omnisci_hashtable.hpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+       });
+     }).wait();
+  }
+
+  std::vector<JoinOneToManyPtrs> lookup(const std::vector<K> &other_keys) {
+    std::vector<JoinOneToManyPtrs> answer(other_keys.size());
+
+    {
+      sycl::buffer<JoinOneToManyPtrs> join(answer);
+      sycl::buffer<K> keys(other_keys);
```
[the remark this answers is not in the document]
Comment (bagrorg) [reply common/dpcpp/omnisci_hashtable.hpp:128]: Apparently, yes
Buffer constructors from pointers and smart pointers allocates memory, so vector is most likely no exception

File: join/join_omnisci.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
+
+#include "join_omnisci.hpp"
+#include "common/dpcpp/omnisci_hashtable.hpp"
+#include "join_helpers/join_helpers.hpp"
+#include <unordered_set>
+
+using JoinOneToManySet = JoinOneToMany<std::unordered_set<size_t>>;
+
+size_t count_distinct(const std::vector<uint32_t> &v) {
```
[the remark this answers is not in the document]
Comment (bagrorg) [reply join/join_omnisci.cpp:9]: We will need to use `std::sort` then, but if we don't want to sort keys then we will copy this vector. What is the best thing to do in this situation then?

**Review by kurapov-peter — APPROVED** [review]

**Pull request merged**